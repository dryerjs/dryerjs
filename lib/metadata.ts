import * as util from './util';

type TargetClass = any;
type MetaValue = any;

const METADATA = Symbol('metadata');
const ENTITY = Symbol('entity');

export enum MetaKey {
  DesignType = 'design:type',
  ParamTypes = 'design:paramtypes',
  ReturnType = 'design:returntype',
  ExcludeOnDatabase = 'ExcludeOnDatabase',
  EmbeddedType = 'EmbeddedType',
  ReferencesManyType = 'ReferencesManyType',
  Thunk = 'Thunk',
  UseProperty = 'UseProperty',
  Entity = 'Entity',
}

export class Metadata {
  public static getPropertiesByModel(
    target: TargetClass,
    metaKey: MetaKey,
  ): { [property: string]: MetaValue } {
    const result = {};
    for (const property of Object.keys(util.defaultTo(target[METADATA], []))) {
      const value = this.getMetaValue(target, metaKey, property);
      if (util.isUndefined(value)) continue;
      result[property] = value;
    }
    return result;
  }

  private static getConstructor(target: TargetClass) {
    return typeof target === 'function' ? target : target.constructor;
  }

  private static setMetaValue(
    target: TargetClass,
    metaKey: MetaKey,
    property: string | symbol,
    value: MetaValue,
  ): void {
    const constructor = this.getConstructor(target);
    if (util.isUndefined(constructor[METADATA])) {
      constructor[METADATA] = {};
    }
    if (util.isUndefined(constructor[METADATA][property])) {
      constructor[METADATA][property] = {};
    }
    constructor[METADATA][property][metaKey] = value;
  }

  private static getMetaValue(target: TargetClass, metaKey: MetaKey, property: string | symbol): MetaValue {
    return this.getConstructor(target)[METADATA]?.[property]?.[metaKey];
  }

  public static for(target: TargetClass) {
    return {
      set(key: MetaKey, value: MetaValue) {
        Metadata.setMetaValue(target, key, ENTITY, value);
      },
      get(key: MetaKey) {
        return Metadata.getMetaValue(target, key, ENTITY);
      },
      with(property: string | symbol) {
        return {
          set(key: MetaKey, value: MetaValue) {
            Metadata.setMetaValue(target, key, property, value);
          },
          get(key: MetaKey) {
            return Metadata.getMetaValue(target, key, property);
          },
        };
      },
    };
  }
}
