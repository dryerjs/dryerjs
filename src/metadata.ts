import 'reflect-metadata';
import * as util from './util';
import {
    EmbeddedSchemaOptions,
    SchemaOptions,
    Relation,
    RelationKind,
    TargetClass,
    FilterableOptions,
} from './shared';

export enum MetaKey {
    DesignType = 'design:type',
    TransformOnOutput = 'TransformOnOutput',
    TransformOnCreate = 'TransformOnCreate',
    TransformOnUpdate = 'TransformOnUpdate',
    Validate = 'Validate',
    DefaultOnCreate = 'DefaultOnCreate',
    DefaultOnUpdate = 'DefaultOnUpdate',
    DefaultOnOutput = 'DefaultOnOutput',
    ExcludeOnOutput = 'ExcludeOnOutput',
    ExcludeOnCreate = 'ExcludeOnCreate',
    ExcludeOnUpdate = 'ExcludeOnUpdate',
    ExcludeOnDatabase = 'ExcludeOnDatabase',
    RequiredOnCreate = 'RequiredOnCreate',
    RequiredOnUpdate = 'RequiredOnUpdate',
    NullableOnOutput = 'NullableOnOutput',
    GraphQLType = 'GraphQLType',
    DatabaseType = 'DatabaseType',
    Enum = 'Enum',
    Embedded = 'Embedded',
    Relation = 'Relation',
    ScalarArrayType = 'ScalarArrayType',
    Schema = 'Schema',
    Filterable = 'Filterable',
}

type AnyEnum = { [key: string]: any };
type MetaValue = any;

export type Ref<T> = T;

const METADATA = Symbol('metadata');
const MODEL_KEY = Symbol('modelKey');

export class Metadata {
    public static getPropertiesByModel(
        target: TargetClass,
        metaKey: MetaKey,
    ): { [property: string]: MetaValue } {
        return util.defaultTo(target[METADATA]?.[metaKey], {});
    }

    public static getMetaValue(target: TargetClass, metaKey: MetaKey, property: string | symbol): MetaValue {
        return target[METADATA]?.[metaKey]?.[property];
    }

    public static addProperty(
        target: TargetClass,
        metaKey: MetaKey,
        property: string | symbol,
        value: MetaValue = true,
    ): void {
        const constructor = typeof target === 'function' ? target : target.constructor;
        if (util.isUndefined(constructor[METADATA])) {
            constructor[METADATA] = {};
        }
        if (util.isUndefined(constructor[METADATA][metaKey])) {
            constructor[METADATA][metaKey] = {};
        }
        constructor[METADATA][metaKey][property] = value;
    }

    public static getModelMetaValue(target: TargetClass, metaKey: MetaKey): MetaValue {
        return this.getMetaValue(target, metaKey, MODEL_KEY);
    }

    public static addModelProperty(target: TargetClass, metaKey: MetaKey, value: MetaValue): void {
        this.addProperty(target, metaKey, MODEL_KEY, value);
    }
}

export function Property(options: { enum?: AnyEnum; type?: TargetClass } = {}) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.DesignType, propertyKey);
        if (util.isNotNullObject(options.enum)) {
            if (Object.keys(options.enum).length !== 1) {
                const message = `Enum should be defined as an object. Example: @Property({ enum: { UserStatus })`;
                throw new Error(message);
            }
            Metadata.addProperty(target, MetaKey.Enum, propertyKey, options.enum);
        }
        if (util.isFunction(options.type)) {
            Metadata.addProperty(target, MetaKey.ScalarArrayType, propertyKey, options.type);
        }
    };
}

export function Schema(options: SchemaOptions) {
    return function (target: TargetClass) {
        Metadata.addModelProperty(target, MetaKey.Schema, options);
    };
}

export function EmbeddedProperty(options: EmbeddedSchemaOptions) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        Metadata.addProperty(target, MetaKey.Embedded, propertyKey, options);
    };
}

type TransformFunction = (value: any, ctx: any, object: any) => any;

export function TransformOnOutput(fn: TransformFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.TransformOnOutput, propertyKey, fn);
    };
}

export function TransformOnInput(fn: TransformFunction) {
    return function (target: TargetClass, propertyKey: string) {
        TransformOnCreate(fn)(target, propertyKey);
        TransformOnUpdate(fn)(target, propertyKey);
    };
}

export function TransformOnCreate(fn: TransformFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.TransformOnCreate, propertyKey, fn);
    };
}

export function TransformOnUpdate(fn: TransformFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.TransformOnUpdate, propertyKey, fn);
    };
}

type DefaultFunction = (ctx: any, object: any) => any;

export function DefaultOnOutput(fn: DefaultFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.DefaultOnOutput, propertyKey, fn);
    };
}

export function DefaultOnInput(fn: DefaultFunction) {
    return function (target: TargetClass, propertyKey: string) {
        DefaultOnCreate(fn)(target, propertyKey);
        DefaultOnUpdate(fn)(target, propertyKey);
    };
}

export function DefaultOnCreate(fn: DefaultFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.DefaultOnCreate, propertyKey, fn);
    };
}

export function DefaultOnUpdate(fn: DefaultFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.DefaultOnUpdate, propertyKey, fn);
    };
}

type ValidateFunction = (value: any, ctx: any, object: any) => any;

export function Validate(fn: ValidateFunction) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.Validate, propertyKey, fn);
    };
}

export function ExcludeOnOutput() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.ExcludeOnOutput, propertyKey);
    };
}

export function ExcludeOnInput() {
    return function (target: TargetClass, propertyKey: string) {
        ExcludeOnCreate()(target, propertyKey);
        ExcludeOnUpdate()(target, propertyKey);
    };
}

export function ExcludeOnCreate() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.ExcludeOnCreate, propertyKey);
    };
}

export function ExcludeOnUpdate() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.ExcludeOnUpdate, propertyKey);
    };
}

export function RequiredOnCreate() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.RequiredOnCreate, propertyKey);
    };
}

export function RequiredOnUpdate() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.RequiredOnUpdate, propertyKey);
    };
}

export function ExcludeOnDatabase() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.ExcludeOnDatabase, propertyKey);
    };
}

export function NullableOnOutput() {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.NullableOnOutput, propertyKey);
    };
}

export function GraphQLType(type: any) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.GraphQLType, propertyKey, type);
    };
}

export function DatabaseType(type: any) {
    return function (target: TargetClass, propertyKey: string) {
        Metadata.addProperty(target, MetaKey.DatabaseType, propertyKey, type);
    };
}

export function HasMany(options: { type: TargetClass; from?: string; to: string }) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        ExcludeOnUpdate()(target, propertyKey);
        ExcludeOnDatabase()(target, propertyKey);
        const relation: Relation = {
            ...options,
            kind: RelationKind.HasMany,
            from: util.defaultTo(options.from, '_id'),
        };
        Metadata.addProperty(target, MetaKey.Relation, propertyKey, relation);
    };
}

export function BelongsTo(options: { type: TargetClass; from: string; to?: string }) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        ExcludeOnInput()(target, propertyKey);
        ExcludeOnDatabase()(target, propertyKey);
        const relation: Relation = {
            ...options,
            kind: RelationKind.BelongsTo,
            to: util.defaultTo(options.to, '_id'),
        };
        Metadata.addProperty(target, MetaKey.Relation, propertyKey, relation);
    };
}

export function HasOne(options: { type: TargetClass; from?: string; to: string }) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        ExcludeOnUpdate()(target, propertyKey);
        ExcludeOnDatabase()(target, propertyKey);
        const relation: Relation = {
            ...options,
            kind: RelationKind.HasOne,
            from: util.defaultTo(options.from, '_id'),
        };
        Metadata.addProperty(target, MetaKey.Relation, propertyKey, relation);
    };
}

export function ReferencesMany(options: { type: TargetClass; from: string; to?: string }) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        ExcludeOnUpdate()(target, propertyKey);
        ExcludeOnDatabase()(target, propertyKey);
        const relation: Relation = {
            ...options,
            kind: RelationKind.ReferencesMany,
            to: util.defaultTo(options.to, '_id'),
        };
        Metadata.addProperty(target, MetaKey.Relation, propertyKey, relation);
    };
}

export function Filterable(options: FilterableOptions) {
    return function (target: TargetClass, propertyKey: string) {
        Property()(target, propertyKey);
        Metadata.addProperty(target, MetaKey.Filterable, propertyKey, options);
    };
}
