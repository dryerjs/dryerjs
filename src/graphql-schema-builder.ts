import {
    GraphQLObjectType,
    GraphQLInputObjectType,
    GraphQLNonNull,
    GraphQLString,
    GraphQLFloat,
    GraphQLBoolean,
    GraphQLEnumType,
} from 'graphql';
import { AnyClass, ModelDefinition } from './type';
import { CachedPropertiesByModel, MetadataKey } from './metadata';

const enumTypeCached = {};

class TraversedProperty {
    constructor(
        private modelDefinition: ModelDefinition,
        public name: string,
        public typeInClass: AnyClass,
    ) {}

    public getMetadataValue(key: MetadataKey) {
        return CachedPropertiesByModel.getMetadataValue(this.modelDefinition.name, key, this.name);
    }
}

abstract class BaseTypeBuilder {
    constructor(protected modelDefinition: ModelDefinition) {}

    private traverse(fn: (property: TraversedProperty) => void) {
        const instance = new this.modelDefinition();
        for (const property in CachedPropertiesByModel.getPropertiesByModel(
            this.modelDefinition.name,
            MetadataKey.DesignType,
        )) {
            const typeInClass = Reflect.getMetadata(MetadataKey.DesignType, instance, property);
            const traversedProperty = new TraversedProperty(this.modelDefinition, property, typeInClass);
            fn(traversedProperty);
        }
    }

    protected abstract getName(): string;
    protected abstract isExcludedField(traversedProperty: TraversedProperty): boolean;
    protected abstract isNullableField(traversedProperty: TraversedProperty): boolean;
    protected abstract useAs: 'input' | 'output';

    public getType() {
        const result = {
            name: this.getName(),
            fields: {},
        };

        this.traverse(traversedProperty => {
            if (this.isExcludedField(traversedProperty)) return;
            const isNullable = this.isNullableField(traversedProperty);
            const type = this.getTypeForOneField(traversedProperty, isNullable);
            if (!type) return;

            result.fields[traversedProperty.name] = { type };
        });

        if (this.useAs === 'input') return new GraphQLInputObjectType(result);
        if (this.useAs === 'output') return new GraphQLObjectType(result);
        /* istanbul ignore next */
        throw new Error('Invalid useAs');
    }

    private getTypeForOneField(traversedProperty: TraversedProperty, nullable: boolean) {
        const baseType = this.getBaseTypeForField(traversedProperty);
        return nullable ? baseType : new GraphQLNonNull(baseType);
    }

    private getBaseTypeForField(traversedProperty: TraversedProperty) {
        const overrideType = traversedProperty.getMetadataValue(MetadataKey.GraphQLType);
        if (overrideType) return overrideType;

        const typeConfig = {
            String: GraphQLString,
            Date: GraphQLString,
            Number: GraphQLFloat,
            Boolean: GraphQLBoolean,
        };
        const enumInObject = traversedProperty.getMetadataValue(MetadataKey.Enum);
        if (enumInObject) {
            const enumName = Object.keys(enumInObject)[0];
            const enumValues = enumInObject[enumName];

            enumTypeCached[enumName] =
                enumTypeCached[enumName] ??
                new GraphQLEnumType({
                    name: enumName,
                    values: Object.keys(enumValues).reduce((values, key) => {
                        values[key] = { value: enumValues[key] };
                        return values;
                    }, {}),
                });

            return enumTypeCached[enumName];
        }

        const scalarBaseType = typeConfig[traversedProperty.typeInClass.name];
        if (scalarBaseType) return scalarBaseType;

        const isEmbedded = traversedProperty.getMetadataValue(MetadataKey.Embedded);
        if (isEmbedded) {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return new this.constructor(traversedProperty.typeInClass).getType();
        }

        throw new Error(
            `Invalid type for field ${traversedProperty.name}. You can override it with @GraphQLType(/* type */)`,
        );
    }
}

class OutputTypeBuilder extends BaseTypeBuilder {
    protected getName() {
        return this.modelDefinition.name;
    }

    protected isExcludedField(traversedProperty: TraversedProperty) {
        return traversedProperty.getMetadataValue(MetadataKey.ExcludeOnOutput);
    }

    protected isNullableField(traversedProperty: TraversedProperty) {
        return traversedProperty.getMetadataValue(MetadataKey.NullableOnOutput);
    }

    protected useAs: 'input' | 'output' = 'output';
}

class CreateInputTypeBuilder extends BaseTypeBuilder {
    protected getName() {
        return `Create${this.modelDefinition.name}Input`;
    }

    protected isExcludedField(traversedProperty: TraversedProperty) {
        return traversedProperty.getMetadataValue(MetadataKey.ExcludeOnCreate);
    }

    protected isNullableField(traversedProperty: TraversedProperty) {
        return !traversedProperty.getMetadataValue(MetadataKey.RequiredOnCreate);
    }

    protected useAs: 'input' | 'output' = 'input';
}

class UpdateInputTypeBuilder extends BaseTypeBuilder {
    protected getName() {
        return `Update${this.modelDefinition.name}Input`;
    }

    protected isExcludedField(traversedProperty: TraversedProperty) {
        return traversedProperty.getMetadataValue(MetadataKey.ExcludeOnUpdate);
    }

    protected isNullableField(traversedProperty: TraversedProperty) {
        return !traversedProperty.getMetadataValue(MetadataKey.RequiredOnUpdate);
    }

    protected useAs: 'input' | 'output' = 'input';
}

export class GraphqlTypeBuilder {
    static build(modelDefinition: ModelDefinition) {
        const output = new OutputTypeBuilder(modelDefinition).getType() as GraphQLObjectType;
        const create = new CreateInputTypeBuilder(modelDefinition).getType() as GraphQLInputObjectType;
        const update = new UpdateInputTypeBuilder(modelDefinition).getType() as GraphQLInputObjectType;
        return {
            output,
            nonNullOutput: new GraphQLNonNull(output),
            create,
            update,
        };
    }
}
