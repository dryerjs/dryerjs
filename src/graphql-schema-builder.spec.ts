import { ExcludeOnDatabase, Property } from './metadata';
import { GraphqlTypeBuilder } from './graphql-schema-builder';

class TestModelDefinition {
    @Property()
    id: string;

    @Property()
    public name: string;

    @ExcludeOnDatabase()
    public excludedField: string;

    @Property()
    public createdAt: Date;
}

describe('GraphqlTypeBuilder', () => {
    it('should build a graphql schema', () => {
        const gqlType = GraphqlTypeBuilder.build(TestModelDefinition);
        expect(gqlType).toMatchSnapshot();
    });
});
