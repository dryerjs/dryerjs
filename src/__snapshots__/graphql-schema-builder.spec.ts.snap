// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphqlTypeBuilder should build a graphql schema 1`] = `
{
  "create": "CreateTestModelDefinitionInput",
  "nonNullOutput": "TestModelDefinition!",
  "output": "TestModelDefinition",
  "paginationOutput": "TestModelDefinitionPagination",
  "update": "UpdateTestModelDefinitionInput",
}
`;
